### DISCUSSION TITLE
Support Prometheus Metrics for Observability

### DISCUSSION THREAD

**Discussion by knechtionscoding** [renovatebot/renovate#24840] [Suggest an Idea]
### What would you like Renovate to be able to do?

As far as I can tell Renovate doesn't support prometheus metrics for consumption into what Renovate is doing. It would be  very useful to be able to measure, how many PRs is it opening, is it failing to open PRs, timeouts or failure to authenticate to github, etc. as a metric so they can be alerted on.

### If you have any ideas on how this should be implemented, please tell us here.

No

### Is this a feature you are interested in implementing yourself?

No

**Comment by rarkins** [comment]
Renovate is not a React app nor a server like Express, so I'm not sure that proposed package would be appropriate. e.g. their Usage is:

```
const apiMetrics = require('prometheus-api-metrics');
app.use(apiMetrics())
```

**Comment by knechtionscoding** [comment]
Ah, well then I'll remove that recommendation, but stand by the feature request.

**Comment by secustor** [comment]
Implementation proposal:
 - Wait for [OTEL MVC](https://github.com/renovatebot/renovate/pull/15731) in v33 
 - Add OTEL metrics implementation 
 - Add [opentelemetry-prometheus-exporter](https://www.npmjs.com/package/@opentelemetry/exporter-prometheus) component and allow to enable it based on environment variables

**Comment by maxbrunet** [comment]
> * Add [opentelemetry-prometheus-exporter](https://www.npmjs.com/package/@opentelemetry/exporter-prometheus) component and allow to enable it based on environment variables

That might be useful for Mend Renovate app (offline processing, I guess), but Renovate OSS runs as a batch job and Prometheus is unlikely to be able to scrape all the metrics before it exits. Pushing metrics to a Prometheus Pushgateway may be more appropriate:

[prometheus.io/docs/practices/instrumentation/#batch-jobs](https://prometheus.io/docs/practices/instrumentation/#batch-jobs)
[prometheus.io/docs/practices/pushing](https://prometheus.io/docs/practices/pushing/)

**Comment by secustor** [comment]
OTEL is push based. 

True, the additional Prometheus exporter is only useful for long running instances.

**Comment by arnouthoebreckx** [comment]
Is there any progress with this request? This feature would be really useful in my current context aswell.

**Comment by movd** [comment]
@secustor Would the renovate project be open for a PR to support this?

Renovate would be ideally suited to push metrics to a [Prometheus Pushgateway](https://prometheus.io/docs/practices/pushing/). 

> The Pushgateway is an intermediary service which allows you to push metrics from jobs which cannot be scraped. 

One use case: We need metrics to check whether Renovate could fetch updates for all repositories. Currently, this is only possible by checking or parsing logs.

There already exists a tool (https://github.com/raffis/renovate-metrics) that does this by reading Renovate's logs, but it would be better to do this inside the application instead of log piping.

**Reply by secustor** [reply]
Yes, we would open to it. 

Now that Prometheus ( with v3 ) natively supports metric ingestion via OTLP, there is not really a reason anymore to send metrics to a PushGateway.

But there needs to be a discussion before about what kind of metrics we are talking about.
